### ISSUE TITLE
Improve handling of snapshot reserve for ontap-nas driver

### ISSUE THREAD

**Issue by wonderland** [NetApp/trident#496]
The current behaviour of the ontap-nas driver is to set the Ontap volume size to the size of the PVC. If a snapshot reserve is configured, the space required for this reserve is taken from the volume size. This leads to a reduced usable/visible size in the pod. For a developer, this is very confusing, as the PVC size does not match the available size in the pod. A developer should not have to know inner workings of Ontap such as snapshot percentage.

**Example**

Take the following simple backend configuration:

```
{
    "version": 1,
    "storageDriverName": "ontap-nas",
    "backendName": "snapreserve",
    "managementLIF": "xxxx",
    "dataLIF": "xxxx",
    "svm": "svm_hland",
    "username": "trident",
    "password": "xxxx",
    "storagePrefix": "trident_",
    "defaults": {
	    "spaceReserve": "none",
	    "exportPolicy": "default",
	    "snapshotReserve":  "50"
    }
}
```

And a matching storage class:

```
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: ontap-nas-reserve
provisioner: csi.trident.netapp.io
parameters:
  backendType: "ontap-nas"
  provisioningType: "thin"
  storagePools: "snapreserve:.*"

```

Now start a simple test pod with a PVC attached:

```
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: testvol
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 5Gi
  storageClassName: ontap-nas-reserve
---
kind: Pod
apiVersion: v1
metadata:
  name: testpod
spec:
  volumes:
    - name: voltest
      persistentVolumeClaim:
       claimName: testvol
  containers:
    - name: ubuntu-test
      image: ubuntu
      command: ["/bin/bash", "-ec", "while :; do echo '.'; sleep 5 ; done"]
      volumeMounts:
        - mountPath: "/my-pv"
          name: voltest
```

When looking at the PVC it has the requested 5GB:

```
:>kubectl get pvc
NAME      STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS        AGE
testvol   Bound    pvc-363a38aa-b9cf-4283-aee3-c3db4b138246   5Gi        RWO            ontap-nas-reserve   2m56s
```

When looking at it inside the pod, only 2.5Gb are visible:

```
:>kubectl exec -it testpod -- /bin/bash
root@testpod:/# df -h /my-pv
Filesystem                                                        Size  Used Avail Use% Mounted on
192.168.184.10:/trident_pvc_363a38aa_b9cf_4283_aee3_c3db4b138246  2.5G  256K  2.5G   1% /my-pv
```

This is very confusing, as PVC size and space available to the pod do not match. Any details around snap reserve are not visible to the developer.

**Why snap reserve is desirable**

Setting the snapshot reserve, and potentially setting it to a rather large value, is beneficial for developer usability. Snapshots can be easily taken through k8s/Trident. While the snapshot doesn't consume any extra space initially, it needs storage capacity as soon as changes to the active file system are made. Without snap reserve, the snapshot will consume capacity from the space visible in the active file system. So when a developer deletes files on the PVC in order to free up space, he will not see any new free space - as any block freed in the active file system are immediately consumed by the snapshot. This can be very confusing unless one is familiar with the inner working of the storage backend. Similar with a small snap reserve: Initially the snapshot consumes capacity from the reserve, so the snapshot does not have any impact on the free space visible in the pod. But once the snapshot(s) outgrows the available capacity in the reserve it again starts to consume capacity from the active file system. A larger snap reserve can prevent this in most cases. Thin provisioning prevents a waste of capacity when working with larger snap reserve.

**Proposal**

Trident should consider the impact of snap reserve when provisoning a new ontap-nas volume. The Ontap volume size should be PVC_size + snap_reserve. So if a PVC of 5GB is requested and snap reserve is configured for 50%, Trident should provision the Ontap volume with 7.5GB.

**Comment by NA-Scott** [comment]
This affects ontap-san as well as ontap-nas.

**Comment by mcpdev80** [comment]
Good morning,

we have the same problem, our developer are confused about this problem.

**Comment by wonderland** [comment]
It actually is a lot worse for ontap-san, imho. With the NAS driver the pod at least sees that the PVC runs out of space. With ontap-san the pod sees free space but all of a sudden cannot write anymore, LUNs go offline etc. I therefore create a dedicated issue for ontap-san: #554

**Comment by gnarl** [comment]
This issue is fixed with commit [e2c03fe](https://github.com/NetApp/trident/commit/e2c03fecd50a7cf92d816708fb407e18120d7981) which will be included in the Trident 21.07 release.
